### PR TITLE
[Fix] TPU Training Type Plugin

### LINKED ISSUE #6390: Fitting hangs at "cleaning up ddp environment..." when tpu_cores=8
## 🐛 Bug

When setting `tpu_cores` of Trainer to 8, fitting hangs at "cleaning up ddp environment...".

## Please reproduce using the BoringModel

https://colab.research.google.com/drive/1tJswNaT0I-GrGsi6ngwwRDUmeFY1pFr3?usp=sharing

### To Reproduce

Run above URL notebook.

### Expected behavior

Trainer.fit ends normally .

### Environment

 - PyTorch Version (e.g., 1.0): 1.7.0a0+7e71a98
 - OS (e.g., Linux): Linux
 - How you installed PyTorch (`conda`, `pip`, source): pip
 - Build command you used (if compiling from source): N/A
 - Python version: 3.7.10
 - CUDA/cuDNN version: N/A
 - GPU models and configuration: N/A
 - Any other relevant information: I tried the notebook on Google Colab with TPU.

### PR DESCRIPTION
**Pull request by kaushikb11**
## What does this PR do?

Fixes #6390
* Fix TPU Colab Hang
* TPU Training Plugin Housekeeping
* Update Accelerator Connector logic for TPUs.

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [ ] Did you make sure to update the documentation with your changes? (if necessary)
- [ ] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [x] Is this pull request ready for review? (if not, please submit in draft mode)
 - [ ] Check that all items from **Before submitting** are resolved
 - [ ] Make sure the title is self-explanatory and the description concisely explains the PR
 - [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
Make sure you had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 1]
```diff
@@ -74,10 +64,9 @@ def distributed_sampler_kwargs(self) -> dict:
     def is_distributed(self):
         return self.world_size != 1
 
-    def process_dataloader(self, dataloader: Union[Iterable, torch.utils.data.DataLoader]) -> ParallelLoader:
+    def process_dataloader(self, dataloader: Union[Iterable, torch.utils.data.DataLoader]) -> MpDeviceLoader:
         device = xm.xla_device()
-        dataloader = xla_pl.ParallelLoader(dataloader, [device])
-        dataloader = dataloader.per_device_loader(device)
+        dataloader = MpDeviceLoader(dataloader, device)
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:69]: Nice !

File: pytorch_lightning/plugins/training_type/tpu_spawn.py [hunk 2]
```diff
@@ -226,39 +178,6 @@ def reduce(self, output, group: Optional[Any] = None, reduce_op: Optional[Union[
 
         return output
 
-    def post_dispatch(self) -> None:
```
Comment (tchaton) [inline pytorch_lightning/plugins/training_type/tpu_spawn.py:229]: You still need to reload the  weights and results from rank 0 to main process.

**Review by tchaton — CHANGES_REQUESTED** [review]
TPUSpawn post_dispath shoudn't be be removed entirely.

File: pytorch_lightning/plugins/training_type/tpu_spawn.py (hunk 2 above)
Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:229]: @tchaton As it is a subclass of `DDPSpawn`, the methods are available.

https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/ddp_spawn.py#L159

Comment (tchaton) [reply pytorch_lightning/plugins/training_type/tpu_spawn.py:229]: Yes, and the mq_queue gets seem in the right order. All good !

**Review by tchaton — APPROVED** [review]

File: tests/models/test_tpu.py [hunk 3]
```diff
@@ -216,7 +216,7 @@ def test_tpu_grad_norm(tmpdir):
     )
 
     model = BoringModel()
-    tpipes.run_model_test(trainer_options, model, on_gpu=False, with_hpc=False)
+    trainer.fit(model)
```
Comment (awaelchli) [inline tests/models/test_tpu.py:219]: so is the accuracy test not desired here?

**Review by awaelchli — APPROVED** [review]

File: tests/models/test_tpu.py (hunk 3 above)
Comment (kaushikb11) [reply tests/models/test_tpu.py]: Yeah, should have it actually. Reverted!

File: pytorch_lightning/plugins/training_type/single_tpu.py [hunk 4]
```diff
@@ -28,17 +23,22 @@
 
 class SingleTPUPlugin(SingleDevicePlugin):
 
-    def __init__(self, device: Union[torch.device, int]):
+    def __init__(self, device: int):
         if isinstance(device, int):
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/single_tpu.py:27]: is there a point to this `if` now that you removed `torch.device`?

File: pytorch_lightning/plugins/training_type/single_tpu.py [hunk 5]
```diff
@@ -28,17 +23,22 @@
...
 
         self.tpu_local_core_rank = 0
         self.tpu_global_core_rank = 0
 
+    @property
     def on_tpu(self) -> bool:
         return True
 
+    @property
+    def is_distributed(self):
```
Comment (carmocca) [inline pytorch_lightning/plugins/training_type/single_tpu.py:39]:
```suggestion
    def is_distributed(self) -> bool:
```

**Review by carmocca — COMMENTED** [review]
Awesome cleanup!

Is there any way we can test the hang is resolved?

File: pytorch_lightning/plugins/training_type/single_tpu.py (hunk 4 above)
Comment (kaushikb11) [reply pytorch_lightning/plugins/training_type/single_tpu.py]: Updated!

**Comment by kaushikb11** [comment]
> Is there any way we can test the hang is resolved?
@carmocca Previously, we used to have additional logic for Kaggle/Colab https://github.com/PyTorchLightning/pytorch-lightning/blob/master/pytorch_lightning/plugins/training_type/tpu_spawn.py#L130

As it is no longer required, the existing tests should be enough to test the training flow.

**Review by carmocca — APPROVED** [review]
